### PR TITLE
[stable8.2] Heal unencrypted file sizes at download time

### PR DESCRIPTION
**Pull request by schiessle**
approved backport of https://github.com/owncloud/core/pull/22579

@PVince81 @MorrisJobke please test and review... Thanks!

### CODE REVIEW

**Comment by schiesbn** [comment]
@jknockaert maybe you can also review the backports.... Thanks!

**Comment by schiesbn** [comment]
failing tests are unrelated, please ignore:

````
08:45:42 Installing ....
08:45:43 ownCloud is not installed - only a limited number of commands are available
08:45:45 
08:45:45 
08:45:45                                  
08:45:45   [OCP\Files\NotFoundException]  
08:45:45                                  
08:45:45 
08:45:45 
08:45:45 maintenance:install [--database="..."] [--database-name="..."] [--database-host="..."] [--database-user="..."] [--database-pass[="..."]] [--database-table-prefix[="..."]] [--admin-user="..."] [--admin-pass="..."] [--data-dir="..."]
08:45:45 
08:45:45 
08:45:45 Kill the swift docker
````

**Comment by PVince81** [comment]
Tested, works :+1:

File: lib/private/files/storage/wrapper/encryption.php [hunk 1]
```diff
@@ -415,6 +419,128 @@ public function fopen($path, $mode) {
...
+				$lastChunkContentEncrypted=substr($lastChunkContentEncrypted, $blockSize);
+			}
+		}
+
+		fclose($stream);
+
+		// we have to decrypt the last chunk to get it actual size
+		$encryptionModule->begin($this->getFullPath($path), $this->uid, 'r', $header, []);
+		$decryptedLastChunk = $encryptionModule->decrypt($lastChunkContentEncrypted, $lastChunkNr . 'end');
+		$decryptedLastChunk .= $encryptionModule->end($this->getFullPath($path), $lastChunkNr . 'end');
```
Comment (jknockaert) [inline lib/private/files/storage/wrapper/encryption.php]: How can this work? `decryp()` and `end()` seem to be defined to take exactly one argument in 8.2 (and 8.1) as well: https://github.com/owncloud/core/blob/fix_broken_unencrypted_size_8.2/apps/encryption/lib/crypto/encryption.php#L318
Has this really been tested in a meaningful way?

Comment (schiesbn) [inline lib/private/files/storage/wrapper/encryption.php]: you are right, this arguments doesn't exists in oC < 9.0... I will remove it

Comment (schiesbn) [inline lib/private/files/storage/wrapper/encryption.php]: @jknockaert fixed... looking forward to your +1 :wink:

**Comment by jknockaert** [comment]
@schiesbn Has this backport actually been properly tested?

**Comment by schiesbn** [comment]
> Has this backport actually been properly tested?

yes, by me and by @PVince81. As said, we didn't noticed the additional parameter because PHP don't care if you call a method with more parameters than expected. It just ignores it.

File: lib/private/files/storage/wrapper/encryption.php [hunk 2]
```diff
@@ -415,6 +419,128 @@ public function fopen($path, $mode) {
...
+
+		$size = $this->storage->filesize($path);
+		$result = $unencryptedSize;
+
+		if ($unencryptedSize < 0 ||
+			($size > 0 && $unencryptedSize === $size)
+		) {
+			// check if we already calculate the unencrypted size for the
+			// given path to avoid recursions
+			if (isset($this->fixUnencryptedSizeOf[$this->getFullPath($path)]) === false) {
```
Comment (jknockaert) [inline lib/private/files/storage/wrapper/encryption.php]: If the analysis https://github.com/owncloud/core/pull/22579#issuecomment-188710468 is right we may be able to do without the recursion test in 8.1 and 8.2, as there are no versions to be determined. Would we want to give it a try just to see?

Comment (schiesbn) [inline lib/private/files/storage/wrapper/encryption.php]: no, people can also write their own encryption modul and we don't want and we can't forbid to use certain file system operation within the encryption module. For now, that's the safe way to do it.

Comment (jknockaert) [inline lib/private/files/storage/wrapper/encryption.php]: OK

**Comment by jknockaert** [comment]
This one is good :+1:

**Pull request merged**